### PR TITLE
Roles in Vault for Nomad so that tokens can be issued by Vault

### LINKED ISSUE #145: Nomad integration with vault (Nomad ACLs)
Steps:

- ~~create default `read` and `write` policies in nomad~~ - taken care of by #146 
- ~~activate vault's nomad backend~~ taken care of by #146 
- create a nomad management token for vault
- add roles in vault so that it can issue tokens for the default `read` and `write` nomad policies

### PR DESCRIPTION
**Pull request by pdmthorsrud**
Leaving this as draft so it can run tests while I write documentation.
Closes #145

### CODE REVIEW

**Comment by pdmthorsrud** [comment]
Please do suggest a better name for the `vault_nomad_integration.yml` file if you have any. I simply couldn't find anything better.

File: template/terraform/main.tf [hunk 1]
```diff
@@ -1,6 +1,22 @@
...
+  token = "master"
+}
+
 provider "nomad" {
   address = "http://127.0.0.1:4646"
+  // The following line is needed if ACLs are enabled in Nomad
+  //  secret_id = "${data.vault_generic_secret.nomad_secret_id.data.secret_id}"
+}
+
+// The following data-source is needed ACLs are enabled in Nomad
```
Comment (pdmthorsrud) [inline template/terraform/main.tf:12]:
```suggestion
// The following data-source is needed if ACLs are enabled in Nomad
```

File: ansible/vault_nomad_integration.yml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  shell:
+    cmd: vault secrets enable nomad
+  environment:
+    VAULT_ADDR: http://127.0.0.1:8200
+    VAULT_TOKEN: master
+
+- name: fetch bootstrap token nomad
+  shell:
+    cmd: vault kv get -field=secret-id secret/nomad-bootstrap-token
+  register: bootstrap
```
Comment (zhenik) [inline ansible/vault_nomad_integration.yml:11]:
```suggestion
  register: bootstrap_token
```

File: ansible/vault_nomad_integration.yml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  shell:
+    cmd: awk '/Secret/ {print $4}' management.token
+  register: secret_id
+
+- name: add nomad management token to vault
+  shell:
+    cmd: vault kv put secret/nomad-management-token accessor-id="{{ accessor_id.stdout }}" secret-id="{{ secret_id.stdout }}"
+  environment:
+    VAULT_ADDR: http://127.0.0.1:8200
+    VAULT_TOKEN: master
```
Comment (zhenik) [inline ansible/vault_nomad_integration.yml:37]: WDYT to put VAULT_TOKEN value `master` to default_variables?

File: template/terraform/main.tf [hunk 4]
```diff
@@ -1,9 +1,25 @@
+provider "vault" {
+  address = "http://127.0.0.1:8200"
+  token = "master"
+}
+
 provider "nomad" {
   address = "http://127.0.0.1:4646"
+  // The following line is needed if ACLs are enabled in Nomad
```
Comment (zhenik) [inline template/terraform/main.tf:8]: We probably need make research about conditional rendering.

**Comment by zhenik** [comment]
> Please do suggest a better name for the `vault_nomad_integration.yml` file if you have any. I simply couldn't find anything better.

I think name is ok. What I wonder is `/ansible` directory structure. I would suggest to have smth like

```
.
│
├── ansible
│   ├── group_vars
│   ├── roles
│   ├── templates
│   ├── other_tasks
│   │   ├── default_vars.yml
│   │   └── nomad_acl_bootstrap.yml
│   ├── ansible.cfg
│   ├── bootstrap.yml
│   ├── install.yml
│   └── hosts
├── ...
```

File: template/terraform/main.tf (hunk 4 above)
Comment (pdmthorsrud) [reply template/terraform/main.tf:8]: Agreed, now that you mention it I have an idea, so I'll tinker with that a little before I call this done

File: ansible/vault_nomad_integration.yml (hunk 3 above)
Comment (pdmthorsrud) [reply ansible/vault_nomad_integration.yml:37]: I've been thinking of the `default_variables`-file as a sort of "These are variables that can be used to alter the system, and there are their default values". The vault token would not be changed on a frequent basis, and not by the user (only those developing the box itself), therefore it doesn't feel right to me putting it in `default_variables`. 

What DOES feel right is maybe having environment variables like NOMAD_TOKEN, VAULT_TOKEN, VAULT_ADDR etc. set when the box starts. Didn't you try and figure something like that out when working out how to pass variables to ansible, and realise there was no easy way?

File: ansible/vault_nomad_integration.yml (hunk 2 above)
Comment (pdmthorsrud) [reply ansible/vault_nomad_integration.yml:11]: That'll require refactoring a few other places as well, but I agree, more verbose. I'll fix it!

Comment (pdmthorsrud) [reply ansible/vault_nomad_integration.yml:11]: Fixed

File: ansible/vault_nomad_integration.yml (hunk 3 above)
Comment (zhenik) [reply ansible/vault_nomad_integration.yml:37]: There is opportunity to pass variables via extra vars. But I think in ur case u can setup env variables, just to append veriable in the .bashrc 
[Example](https://github.com/fredrikhgrelland/vagrant-hashistack/blob/master/ansible/install.yml#L125-L131)

Comment (pdmthorsrud) [reply ansible/vault_nomad_integration.yml:37]: Right, that's a quick and easy way. I'll get on that

Comment (pdmthorsrud) [reply ansible/vault_nomad_integration.yml:37]: Update: this should be changed, but I'll bundle that with another PR instead

File: template/terraform/main.tf (hunk 4 above)
Comment (pdmthorsrud) [reply template/terraform/main.tf:8]: Forgot to say, but same as above, this should be updated, but will create an issue and bundle it in another PR.

**Pull request merged**